### PR TITLE
Validate org-user membership from gateway

### PR DESCRIPTION
**Pull request by fosterseth**
##### SUMMARY

Adding credential and execution environment roles validates that the user belongs to the same org as the credential or EE.

In some situations, the user-org membership has not yet been synced from gateway to controller.

In this case, controller will make a request to gateway to check if the user is part of the org.

relies on https://github.com/ansible/aap-gateway/pull/557
and
https://github.com/ansible/django-ansible-base/pull/614

##### ISSUE TYPE

 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME

 - API

### CODE REVIEW

File: awx/main/models/credential.py
Comment (AlanCoding) [inline awx/main/models/credential.py]: good to rename gateway -> resource_server

File: awx/main/models/credential.py
Comment (AlanCoding) [inline awx/main/models/credential.py:91]: In this case, we should use the existing check `actor.has_obj_perm(self.organization, 'view')` so that standalone remains functional.

Might do best to move that out of this function, your call.

File: awx/main/models/credential.py
Comment (pb82) [inline awx/main/models/credential.py]: Is `RESOURCE_SERVICE_PATH` a new setting? Do we need installer changes to add it?

Comment (fosterseth) [reply awx/main/models/credential.py]: it's newish but should have a default assigned

File: awx/main/models/credential.py
Comment (fosterseth) [reply awx/main/models/credential.py:91]: i'm still checking if user belongs to org (locally) in the validate methods, that way it short circuits and we don't even get to this function

File: awx/main/models/credential.py
Comment (fosterseth) [reply awx/main/models/credential.py]: changed

File: requirements/requirements_git.txt
Comment (fosterseth) [inline requirements/requirements_git.txt]: change this back before merging

**Comment by AlanCoding** [comment]
Reproduced issue:
 - create new organization in UI
 - created new user in UI and put in that organization
 - created new credential in that organization
 - give user access to that credential

this gives a 400, so I will try to re-do using your patches.

**Comment by AlanCoding** [comment]
Tried them together and got this:

```
T /api/v2/me/ => generated 1051 bytes in 21 msecs (HTTP/1.1 200) 14 headers in 458 bytes (1 switches on core 0)
2024-09-13 19:05:09,604 ERROR    [d8e59a2c] django.request Internal Server Error: /api/v2/role_user_assignments/
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/viewsets.py", line 124, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/awx_devel/awx/api/generics.py", line 351, in dispatch
    return super(APIView, self).dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/rbac/api/views.py", line 133, in perform_create
    return super().perform_create(serializer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/mixins.py", line 24, in perform_create
    serializer.save()
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/lib/serializers/validation.py", line 26, in save
    return super().save(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/serializers.py", line 208, in save
    self.instance = self.create(validated_data)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/rbac/api/serializers.py", line 259, in create
    obj.validate_role_assignment(actor, rd, requesting_user=requesting_user)
  File "/awx_devel/awx/main/models/credential/__init__.py", line 366, in validate_role_assignment
    if check_resource_server_for_user_in_organization(actor, self.organization, requesting_user):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/awx_devel/awx/main/models/credential/__init__.py", line 104, in check_resource_server_for_user_in_organization
    response = client._make_request(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/resource_registry/rest_client.py", line 114, in _make_request
    resp.raise_for_status()
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://aap-gw-proxy-1:9080/api/gateway/v1/role_user_assignments/?role_definition__name__in=Organization%20Member,Organization%20Admin&user__resource__ansible_id=1c36c6ec-c5dc-4230-860a-2069446bf662&object_id=3
2024-09-13 19:05:09,604 ERROR    [d8e59a2c] django.request Internal Server Error: /api/v2/role_user_assignments/
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/viewsets.py", line 124, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/awx_devel/awx/api/generics.py", line 351, in dispatch
    return super(APIView, self).dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/mixins.py", line 19, in create
    self.perform_create(serializer)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/rbac/api/views.py", line 133, in perform_create
    return super().perform_create(serializer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/mixins.py", line 24, in perform_create
    serializer.save()
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/lib/serializers/validation.py", line 26, in save
    return super().save(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/rest_framework/serializers.py", line 208, in save
    self.instance = self.create(validated_data)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/rbac/api/serializers.py", line 259, in create
    obj.validate_role_assignment(actor, rd, requesting_user=requesting_user)
  File "/awx_devel/awx/main/models/credential/__init__.py", line 366, in validate_role_assignment
    if check_resource_server_for_user_in_organization(actor, self.organization, requesting_user):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/awx_devel/awx/main/models/credential/__init__.py", line 104, in check_resource_server_for_user_in_organization
    response = client._make_request(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/ansible_base/resource_registry/rest_client.py", line 114, in _make_request
    resp.raise_for_status()
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://aap-gw-proxy-1:9080/api/gateway/v1/role_user_assignments/?role_definition__name__in=Organization%20Member,Organization%20Admin&user__resource__ansible_id=1c36c6ec-c5dc-4230-860a-2069446bf662&object_id=3
```

We should seriously consider catching this type of error in some way.

**Comment by AlanCoding** [comment]
Wait, forgot one patch

**Comment by AlanCoding** [comment]
Screenshot from 2024-09-13 15-07-50

Ok now it worked.

**Review by AlanCoding — APPROVED** [review]
Obviously be very careful about merge order, but I confirmed the fix in aap-dev

**Comment by webknjaz** [comment]
plz rebase to get coverage uploaded

**Comment by fosterseth** [comment]
note: if requests to gateway are 400 errors, we logger the error
because they are in credential.py file, the log message prefixes with `awx.main.credential`, even in the case of execution environment related error. We could improve this by moving the function outside of credential.py, but really this entire method will eventually go away so not concerned.

**Pull request merged**
